File: app/src/main/java/com/syroniko/casseteapp/LogInSignUp/CountrySelectSignUpActivity.java
package com.syroniko.casseteapp.LogInSignUp;

import android.content.Intent;
import android.graphics.Typeface;
import android.util.Log;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.FirebaseFirestore;
import com.syroniko.casseteapp.R;
import com.syroniko.casseteapp.MainClasses.User;
import com.syroniko.casseteapp.firebase.Auth;
import com.syroniko.casseteapp.firebase.AuthCallback;
import com.syroniko.casseteapp.firebase.UserDB;
import com.ybs.countrypicker.CountryPicker;
import com.ybs.countrypicker.CountryPickerListener;

import org.jetbrains.annotations.Nullable;

public class CountrySelectSignUpActivity extends AppCompatActivity {
    public static final String COUNTRY_USER_EXTRA = "genres user extra";
    public static final String COUNTRY_PASSWORD_EXTRA = "genres password extra";

    private static final String TAG = CountrySelectSignUpActivity.class.getSimpleName();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_country_select_sign_up);

        TextView locationTextView = findViewById(R.id.location_text_view);
        TextView finishTextView = findViewById(R.id.finsihTextView);
        final TextView countryTextView = findViewById(R.id.countryTextView);

        TextView tx = findViewById(R.id.pick_country_string);
        Typeface custom_font = Typeface.createFromAsset(getAssets(),  "fonts/montsextrathic.ttf");
        tx.setTypeface(custom_font);

        Intent intent = getIntent();
        final User user = intent.getParcelableExtra(COUNTRY_USER_EXTRA);
        final String password = intent.getStringExtra(COUNTRY_PASSWORD_EXTRA);


        locationTextView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                final CountryPicker picker = CountryPicker.newInstance("Select Country");
                picker.setListener(new CountryPickerListener() {
                    @Override
                    public void onSelectCountry(String name, String code, String dialCode, int flagDrawableResID) {
                        countryTextView.setText(name);
                        user.setCountry(name);
                        picker.dismiss();
                    }
                });
                picker.show(getSupportFragmentManager(), "COUNTRY_PICKER");
            }
        });

        finishTextView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (user.getCountry() != null) {
//                    FirebaseAuth auth = FirebaseAuth.getInstance();
//
//                    auth.createUserWithEmailAndPassword(user.getEmail(), password)
//                            .addOnCompleteListener(new OnCompleteListener<AuthResult>() {
//                                @Override
//                                public void onComplete(@NonNull Task<AuthResult> task) {
//                                    Toast.makeText(CountrySelectSignUpActivity.this, "Sign Up Comleted", Toast.LENGTH_SHORT).show();
//
//                                    FirebaseFirestore firebaseFirestore = FirebaseFirestore.getInstance();
//                                    String uid = FirebaseAuth.getInstance().getUid();
//                                    if(uid != null) {
//                                        user.setUid(uid);
//                                        firebaseFirestore.collection("users").document(uid).set(user);
//                                    }
//
//                                    //TODO add a cover here
//
//                                    Intent resultIntent = new Intent();
//                                    setResult(RESULT_OK, resultIntent);
//                                    finish();
//                                }
//                            })
//                            .addOnFailureListener(new OnFailureListener() {
//                                @Override
//                                public void onFailure(@NonNull Exception e) {
//                                    Toast.makeText(CountrySelectSignUpActivity.this, "Sign Up Failed. Please try again.", Toast.LENGTH_SHORT).show();
//                                    Log.e(TAG, e.toString());
//                                }
//                            });

                    Auth.INSTANCE.signUpWithEmail(
                            user.getEmail(),
                            password,
                            new AuthCallback() {
                                @Override
                                public void onSuccess(@Nullable String uid) {
                                    Toast.makeText(CountrySelectSignUpActivity.this, "Sign Up Completed", Toast.LENGTH_SHORT).show();
                                    if(uid != null) {
                                        user.setUid(uid);
                                        UserDB.INSTANCE.addUser(uid, user);

                                        Intent resultIntent = new Intent();
                                        setResult(RESULT_OK, resultIntent);
                                        finish();
                                    }
                                }

                                @Override
                                public void onFailure() {
                                    Toast.makeText(CountrySelectSignUpActivity.this, "Sign Up Failed. Please try again.", Toast.LENGTH_SHORT).show();
                                }
                            }
                    );
                }
                else{
                    Toast.makeText(CountrySelectSignUpActivity.this, "Please select a country.", Toast.LENGTH_SHORT).show();
                }
            }
        });
    }

}

File: app/src/main/java/com/syroniko/casseteapp/LogInSignUp/CountrySelectSignUpActivity.kt
package com.syroniko.casseteapp.LogInSignUp

import android.content.Context
import android.content.Intent
import android.graphics.Typeface
import android.os.Bundle
import android.util.Log
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.syroniko.casseteapp.MainClasses.MainActivity
import com.syroniko.casseteapp.MainClasses.User
import com.syroniko.casseteapp.MainClasses.toast
import com.syroniko.casseteapp.R
import com.syroniko.casseteapp.utils.SPOTIFY_NO_TOKEN
import com.syroniko.casseteapp.firebase.Auth
import com.syroniko.casseteapp.firebase.AuthCallback
import com.syroniko.casseteapp.firebase.UserDB
import com.ybs.countrypicker.CountryPicker
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class CountrySelectSignUpActivity : AppCompatActivity() {

    //todo add auth functionality to a view model
    private val auth = Auth()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_country_select_sign_up)

        val locationTextView = findViewById<TextView>(R.id.location_text_view)
        val finishTextView = findViewById<TextView>(R.id.finsihTextView)
        val countryTextView = findViewById<TextView>(R.id.countryTextView)
        val tx = findViewById<TextView>(R.id.pick_country_string)
        val customFont = Typeface.createFromAsset(assets, "fonts/montsextrathic.ttf")
        tx.typeface = customFont

        val user = intent.getParcelableExtra<User>(COUNTRY_USER_EXTRA) ?: return
        val password = intent.getStringExtra(COUNTRY_PASSWORD_EXTRA) ?: return
        val spotifyToken = intent.getStringExtra(SPOTIFY_TOKEN_EXTRA_NAME) ?: SPOTIFY_NO_TOKEN

        locationTextView.setOnClickListener {
            val picker = CountryPicker.newInstance("Select Country")

            picker.setListener { name, _, _, _ ->
                countryTextView.text = name
                user.country = name
                picker.dismiss()
            }

            picker.show(supportFragmentManager, "COUNTRY_PICKER")
        }

        finishTextView.setOnClickListener {
            if (user.country != null) {
                auth.signUpWithEmail(
                    user.email,
                    password,
                    object : AuthCallback {
                        override fun onSuccess(uid: String?) {
                            toast("Sign Up Completed.")

                            if (uid != null) {
                                Log.d(CountrySelectSignUpActivity::class.java.simpleName, "123")
                                user.uid = uid
                                Log.d(CountrySelectSignUpActivity::class.java.simpleName, uid)
                                UserDB.insert(user)
////                                TODO("What is that? :O")
//                                val resultIntent = Intent()
//                                setResult(Activity.RESULT_OK, resultIntent)
//                                finish()
                                MainActivity.startActivity(this@CountrySelectSignUpActivity, uid, user, spotifyToken)
                            }

                        }

                        override fun onFailure() {
                            toast("Sign Up Failed. Please try again.")
                        }
                    }
                )

            }
            else {
                toast("Please select a country.")
            }
        }
    }

    companion object {
        private const val COUNTRY_USER_EXTRA = "genres user extra"
        private const val COUNTRY_PASSWORD_EXTRA = "genres password extra"
        private const val SPOTIFY_TOKEN_EXTRA_NAME = "spotify token extra name"

        fun startActivity(context: Context, user: User, password: String, spotifyToken: String){
            val intent = Intent(context, CountrySelectSignUpActivity::class.java)
            intent.putExtra(COUNTRY_USER_EXTRA, user)
            intent.putExtra(COUNTRY_PASSWORD_EXTRA, password)
            intent.putExtra(SPOTIFY_TOKEN_EXTRA_NAME, spotifyToken)
            intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP
            context.startActivity(intent)
        }
    }
}